File: 907-koko-eating-bananas/koko-eating-bananas.go
func minEatingSpeed(p []int, h int) int {
	sort.Ints(p)
	l, r := 1, p[len(p)-1]

	for l <= r {
		mid := l + (r-l)/2
		c := 0

		for i := 0; i < len(p); i++ {
			c += (p[i] + mid - 1) / mid
		}

		if c > h {
			l = mid + 1
		} else {
			r = mid - 1
		}
	}

	return l
}

File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& p, int h) {
        sort(p.begin(), p.end());
        int l = 1; 
        int r = p.back();
        
        while (l <= r) {
            int mid = l + (r - l) / 2;
            long long hours = 0;
            
            for (int i = 0; i < p.size(); ++i) {
                hours += (p[i] + mid - 1) / mid; 
            }
            
            if (hours > h) {
                l = mid + 1;
            } else {
                r = mid - 1; 
            }
        }
        
        return l;
    }
};
